packages/go_router/lib/src/delegate.dart

```diff
@@ -146,6 +146,18 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
     notifyListeners();
   }
 
+  /// Calls [pop] repeatedly until the predicate returns true.
+  void popUntil(bool Function(RouteMatch) predicate) {
```

Comment:
I think you can just use `go` to build up to a certain routematch list? If we do refactor the imperative API this will be obsolete

Reply:
no I meant it seems like the `go` can replace `popUntil`. Since developer creates their own route table, they will be able to use `go` to achieve functional equivalence  of `popUntil`. For example if they have route table like this
```
GoRouteA(
  routes: [
    GoRouteB(
      routes: [GoRouteC()]
    )
  ],
)
```

lets say route match list is `[GoRouteA, GoRouteB, GoRouteC]`.
`popUntil((route) => route == GoRouteB)` will be the same as `go(GoRouteB)`

Reply:
I think this comes back to https://github.com/flutter/flutter/issues/99112 's discussion.

Here you are assuming that the match list is `[GoRouteA, GoRouteB, GoRouteC]` and matches perfectly the structure
```
GoRouteA(
  routes: [
    GoRouteB(
      routes: [GoRouteC()]
    )
  ],
)
```
And `pop` from `GoRouteC` always goes _"back"_ to `GoRouteB`.

But as I described in https://github.com/flutter/flutter/issues/99112#issuecomment-1301628808, this is not what users are usually expecting. They don't always access `GoRouteC` from `GoRouteB`. So always _"going back"_ to `GoRouteB` from `GoRouteC` doesn't work.
The way I understand it is that you are trying to merge/combine "app location" and "history"? But that's 2 different things that should be independent.

Am I missing something?

Reply:
no, you are right, the previous comment is under the assuming that the match list matches the structure. I am assuming `history` means the internal route stack, not the browser history. In that case, my proposal user can still modify the history through imperative API, it just that they won't be able to change app location with imperative API. Let's continue the discussion in the issue.